File: src/test/java/workshop/part2a/VettingSupervisorTest.java
package workshop.part2a;

import akka.actor.ActorContext;
import akka.actor.Props;
import akka.actor.SupervisorStrategy;
import akka.testkit.TestActorRef;
import akka.testkit.TestProbe;
import org.junit.Test;
import workshop.common.ad.Ad;
import workshop.part1.AkkaTest;
import workshop.part1.Verdict.VerdictType;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.Is.is;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.isA;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class VettingSupervisorTest extends AkkaTest {

    @Test
    public void createsVettingActorWhenAdIsReceived () {
        VettingActorFactory vettingActorFactory = mock(VettingActorFactory.class);

        TestProbe vettingActor = TestProbe.apply(system);
        when(vettingActorFactory.create(any()))
            .thenReturn(vettingActor.ref());

        Ad ad = createAd(123);
        createVettingSupervisor(vettingActorFactory).tell(ad, sender.ref());

        verify(vettingActorFactory).create(isA(ActorContext.class));
        assertThat(vettingActor.expectMsgClass(Ad.class), is(ad));
    }

    @Test
    public void spoofsSenderWhenCreatingVettingActor() {
        VettingActorFactory vettingActorFactory = mock(VettingActorFactory.class);

        TestProbe vettingActor = TestProbe.apply(system);
        when(vettingActorFactory.create(any()))
            .thenReturn(vettingActor.ref());

        Ad ad = createAd(123);
        createVettingSupervisor(vettingActorFactory).tell(ad, sender.ref());

        vettingActor.expectMsgClass(Ad.class);
        vettingActor.reply(VerdictType.GOOD);
        assertThat(sender.expectMsgClass(VerdictType.class), is(VerdictType.GOOD));
    }

    @Test
    public void resumesVettingActorWhenUserNotFoundException() {
        SupervisorStrategy.Directive strategy = createVettingSupervisor(mock(VettingActorFactory.class))
            .underlyingActor()
            .supervisorStrategy()
            .decider()
            .apply(new UserNotFoundException("user not found test exception"));

        assertThat(strategy, is(SupervisorStrategy.resume()));
    }

    @Test
    public void restartsVettingActorWhenNullPointerException() {
        SupervisorStrategy.Directive strategy = createVettingSupervisor(mock(VettingActorFactory.class))
            .underlyingActor()
            .supervisorStrategy()
            .decider()
            .apply(new NullPointerException("null pointer test exception"));

        assertThat(strategy, is(SupervisorStrategy.restart()));
    }

    @Test
    public void escalatesExceptionWhenVettingActorFailsWithOtherException() {
        SupervisorStrategy.Directive strategy = createVettingSupervisor(mock(VettingActorFactory.class))
            .underlyingActor()
            .supervisorStrategy()
            .decider()
            .apply(new RuntimeException("other exception"));

        assertThat(strategy, is(SupervisorStrategy.escalate()));
    }

    private TestActorRef<VettingSupervisor> createVettingSupervisor(VettingActorFactory vettingActorFactory) {
        return TestActorRef.create(system, Props.create(VettingSupervisor.class, () -> new VettingSupervisor(vettingActorFactory)));
    }

    private Ad createAd(int userId) {
        return new Ad(userId, "ad title", "description");
    }
}

File: src/test/kotlin/workshop/part2a/VettingSupervisorTest.kt
package workshop.part2a

import akka.actor.Props
import akka.actor.SupervisorStrategy
import akka.testkit.TestActorRef
import akka.testkit.TestProbe
import com.natpryce.hamkrest.assertion.assertThat
import com.natpryce.hamkrest.equalTo
import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import org.junit.Test
import org.mockito.Mockito.mock
import workshop.common.ad.Ad
import workshop.part1.AkkaTest
import workshop.part1.Verdict


class VettingSupervisorTest : AkkaTest() {

    @Test
    fun createsVettingActorWhenAdIsReceived() {
        val vettingActorFactory = mock(VettingActorFactory::class.java)

        whenever(vettingActorFactory.create(any()))
                .thenReturn(sender.ref())

        val ad = createAd(123)
        createVettingSupervisor(vettingActorFactory).tell(ad, sender.ref())

        verify(vettingActorFactory).create(any())
        assertThat(sender.expectMsgClass(Ad::class.java), equalTo(ad))
    }

    @Test
    fun spoofsSenderWhenCreatingVettingActor() {
        val vettingActorFactory = mock(VettingActorFactory::class.java)

        val vettingActor = TestProbe.apply(system)
        whenever(vettingActorFactory.create(any()))
                .thenReturn(vettingActor.ref())

        val ad = createAd(123)
        createVettingSupervisor(vettingActorFactory).tell(ad, sender.ref())

        vettingActor.expectMsgClass(Ad::class.java)
        vettingActor.reply(Verdict.GOOD)
        assertThat(sender.expectMsgClass(Verdict::class.java), equalTo(Verdict.GOOD))
    }

    @Test
    fun resumesVettingActorWhenUserNotFoundException() {
        val strategy = createVettingSupervisor(mock(VettingActorFactory::class.java))
                .underlyingActor()
                .supervisorStrategy()
                .decider()
                .apply(UserNotFoundException("user not found exception"))

        assertThat(strategy, equalTo(SupervisorStrategy.resume() as SupervisorStrategy.Directive))
    }

    @Test
    fun restartsVettingActorWhenNullPointerException() {
        val strategy = createVettingSupervisor(mock(VettingActorFactory::class.java))
                .underlyingActor()
                .supervisorStrategy()
                .decider()
                .apply(NullPointerException("null pointer exception"))

        assertThat(strategy, equalTo(SupervisorStrategy.restart() as SupervisorStrategy.Directive))
    }

    @Test
    fun escalatesExceptionWhenVettingActorFailsWithOtherException() {
        val strategy = createVettingSupervisor(mock(VettingActorFactory::class.java))
                .underlyingActor()
                .supervisorStrategy()
                .decider()
                .apply(RuntimeException("other exception"))

        assertThat(strategy, equalTo(SupervisorStrategy.escalate() as SupervisorStrategy.Directive))
    }

    private fun createVettingSupervisor(vettingActorFactory: VettingActorFactory): TestActorRef<VettingSupervisor> {
        return TestActorRef.create(system, Props.create(VettingSupervisor::class.java) { VettingSupervisor(vettingActorFactory) })
    }

    private fun createAd(userId: Int): Ad {
        return Ad(userId, "ad title", "description")
    }
}